### PR TITLE
WIP: kernel map iterator to evict stale flows and avoid hmap churns

### PR DESCRIPTION
**Pull request by msherif1234**
## Dependencies

n/a

## Checklist

If you are not familiar with our processes or don't know what to answer in the list below, let us know in a comment: the maintainers will take care of that.

* [ ] Will this change affect NetObserv / Network Observability operator? If not, you can ignore the rest of this checklist.
* [ ] Is this PR backed with a JIRA ticket? If so, make sure it is written as a title prefix _(in general, PRs affecting the NetObserv/Network Observability product should be backed with a JIRA ticket - especially if they bring user facing changes)._
* [ ] Does this PR require product documentation?
  * [ ] If so, make sure the JIRA epic is labelled with "documentation" and provides a description relevant for doc writers, such as use cases or scenarios. Any required step to activate or configure the feature should be documented there, such as new CRD knobs.
* [ ] Does this PR require a product release notes entry?
  * [ ] If so, fill in "Release Note Text" in the JIRA.
* [ ] Is there anything else the QE team should know before testing? E.g: configuration changes, environment setup, etc.
  * [ ] If so, make sure it is described in the JIRA ticket.
* QE requirements (check 1 from the list):
  * [ ] Standard QE validation, with pre-merge tests unless stated otherwise.
  * [ ] Regression tests only (e.g. refactoring with no user-facing change).
  * [ ] No QE (e.g. trivial change with high reviewer's confidence, or per agreement with the QE team).

### CODE REVIEW

File: bpf/flows.c [hunk 1]
```diff
@@ -61,6 +66,10 @@ static inline void update_existing_flow(flow_metrics *aggregate_flow, pkt_info *
     if (aggregate_flow->start_mono_time_ts == 0) {
         aggregate_flow->start_mono_time_ts = pkt->current_ts;
     }
+
+    if (aggregate_flow->last_seen_timestamp == 0) {
```
Comment (tohojo) [inline bpf/flows.c:70]: This should be `if (pkt->currents_ts != 0)` - i.e., always update the timestamp with the latest value.

File: pkg/tracer/tracer.go [hunk 2]
```diff
@@ -756,6 +769,65 @@ func (m *FlowFetcher) LookupAndDeleteMap(met *metrics.Metrics) map[ebpf.BpfFlowI
...
+	}
+
+	// trigger iterator kernel check to look for stale entries
+	f, err := m.flowMapLink.Open()
+	if err != nil {
+		log.WithError(err).WithField("flowId", id).Warnf("couldn't open flows map iterator")
+		m.lookupAndDeleteIterateSupported = false
+		return m.lookupAndDeleteMapUserSpace(met)
+	}
+	f.Close()
```
Comment (tohojo) [inline pkg/tracer/tracer.go:823]: It's not enough to open and close the link, you also need to read from it. See: 

https://github.com/xdp-project/bpf-examples/blob/master/pping/pping.c#L696

**Review by tohojo — COMMENTED** [review]
A couple of nits on the usage of the iterator. But as a wider point, I'm not sure if this approach will give you the right values. Flows that don't expire will have their full counters read on every poll, so you'll get aggregate counter values instead of the values since last poll. So the aggregation logic will have to take this into account, and if you end up calculating the delta in userspace afterwards, I think that will eat any gains there may be from this...

File: pkg/tracer/tracer.go (hunk 2 above)
Comment (msherif1234) [reply pkg/tracer/tracer.go:823]: ok I know I was missing something with iterator userspace trigger

**Comment by msherif1234** [comment]
image

**Comment by msherif1234** [comment]
this approach doesn't seem to yield in good resources usage vs doing it all in userspace
image

**Comment by msherif1234** [comment]
closing this PR as it turned out it won't really help much the userspace resources

**Pull request closed without merging**